# Vyprávíme příběh s daty

## Seznámení s daty

## Pandas 🐼

### Nasťa a Filip Sedlákovi

## Seznámení s daty - postup

1. Načíst data
    - Případně přejmenovat sloupce

2. V jakém formátu jsou sloupce?

3. Kolik je řádků a sloupců?

4. Obsahují některé sloupce chybějící data?

5. Udělat popisnou statistiku dat

## Co je pandas?

- Knihovna pro analýzu dat v Pythonu
- Je postavena na NumPy (knihovna pro práci s numerickými daty)
- [10 minutes to pandas](https://pandas.pydata.org/pandas-docs/stable/10min.html)
- [Srovnání Pandas s SQL](https://pandas.pydata.org/pandas-docs/stable/getting_started/comparison/comparison_with_sql.html)


## Jak naimportuji pandas?

In [1]:
# Importujeme pandas jako zkrátku
# Není to nutné, ale většina tutorialu to tak má
import pandas as pd

## Jak načtu data?

Pandas má několik funkcí, tady je [hezký](https://pandas.pydata.org/pandas-docs/stable/cookbook.html#data-in-out) přehled

Nejpoužívanější je funkce [`read_csv`](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.read_csv.html), která má více než 20 parametrů (nebojte se, jenom jeden parametr je povinný 🙂)

In [2]:
data = pd.read_csv('http://vincentarelbundock.github.io/Rdatasets/csv/DAAG/primates.csv')


In [3]:
# Zobrazir první 2 řádky
data.head(n=2)

Unnamed: 0  Bodywt  Brainwt
0  Potar monkey    10.0      115
1       Gorilla   207.0      406

In [4]:
# Zapsát náhodné 3 řádky do CSV souboru
data1 = data.sample(n=3)
data1.to_csv('data1.csv')

In [5]:
# Zjistit datový typ tabulky
type(data)

pandas.core.frame.DataFrame

In [6]:
# Zjistit datový typ sloupce
type(data.Bodywt)

pandas.core.series.Series

In [7]:
primates = pd.read_csv('primates.csv')
primates.head()

Unnamed: 0  Bodywt  Brainwt
0   Potar monkey    10.0      115
1        Gorilla   207.0      406
2          Human    62.0     1320
3  Rhesus monkey     6.8      179
4          Chimp    52.2      440

In [8]:
# Zobrazíme index tabulky
primates.index

RangeIndex(start=0, stop=5, step=1)

In [9]:
# Zobrazíme názvy sloupců tabulky
data.columns

Index(['Unnamed: 0', 'Bodywt', 'Brainwt'], dtype='object')

## Cvičení
 
Načtěte data `tips.csv` (jsou ve stejné složce) a použijte první sloupec jako index

**Bonus**: Načtete [data z UNICEF](https://data.unicef.org/wp-content/uploads/2015/12/Out-of-School-Rate-Final-for-website.xlsx) pomocí `read_excel` funkce z pandas a přeskočte prvních 9 řádků

In [10]:
tips = pd.read_csv('tips.csv')
tips.head()

Unnamed: 0  total_bill   tip     sex smoker  day    time  size
0           1       16.99  1.01  Female     No  Sun  Dinner     2
1           2       10.34  1.66    Male     No  Sun  Dinner     3
2           3       21.01  3.50    Male     No  Sun  Dinner     3
3           4       23.68  3.31    Male     No  Sun  Dinner     2
4           5       24.59  3.61  Female     No  Sun  Dinner     4

In [11]:
tips = pd.read_csv('tips.csv', index_col=0)
tips.head()

total_bill   tip     sex smoker  day    time  size
1       16.99  1.01  Female     No  Sun  Dinner     2
2       10.34  1.66    Male     No  Sun  Dinner     3
3       21.01  3.50    Male     No  Sun  Dinner     3
4       23.68  3.31    Male     No  Sun  Dinner     2
5       24.59  3.61  Female     No  Sun  Dinner     4

In [12]:
# Bonus načtení Excelu
unicef = pd.read_excel('Out-of-School-Rate-Final-for-website.xlsx', 
                       skiprows=9
)
unicef.head()

ISO3 Code    Countries    Total Unnamed: 3     Male Unnamed: 5   Female  \
0       AFG  Afghanistan     36.3        NaN     26.9        NaN     46.8   
1       ALB      Albania  7.60232          x  7.76783          x  7.41396   
2       DZA      Algeria  2.19084        NaN   2.1955        NaN  2.18592   
3       AND      Andorra      NaN        NaN      NaN        NaN      NaN   
4       AGO       Angola     23.7          y     22.8          y     24.6   

  Unnamed: 7    Urban Unnamed: 9     ...       Poorer Unnamed: 15  Middle  \
0        NaN     19.7        NaN     ...         46.2         NaN    44.4   
1          x  6.66002          x     ...      6.45801           x  8.8039   
2        NaN  1.94815        NaN     ...      2.57707         NaN  1.7089   
3        NaN      NaN        NaN     ...          NaN         NaN     NaN   
4          y     15.1          y     ...         25.8           y    22.5   

  Unnamed: 17   Fourth Unnamed: 19  Richest Unnamed: 21        Source  \
0         NaN     31.8         NaN     16.7         NaN      DHS 2015   
1           x   6.7463           x  6.39511           x   DHS 2008-09   
2         NaN  1.65181         NaN  1.42334         NaN  MICS 2012-13   
3         NaN      NaN         NaN      NaN         NaN           NaN   
4           y     14.4           y      9.7           y  IBEP 2008-09   

  Unnamed: 23  
0         NaN  
1         NaN  
2         NaN  
3         NaN  
4         NaN  

[5 rows x 24 columns]

### Jak můžu odstranít sloupce bez názvu?

In [13]:
# Zobrazíme všechny sloupce
unicef.columns

Index(['ISO3 Code', 'Countries', 'Total', 'Unnamed: 3', 'Male', 'Unnamed: 5',
       'Female', 'Unnamed: 7', 'Urban', 'Unnamed: 9', 'Rural', 'Unnamed: 11',
       'Poorest', 'Unnamed: 13', 'Poorer', 'Unnamed: 15', 'Middle',
       'Unnamed: 17', 'Fourth', 'Unnamed: 19', 'Richest', 'Unnamed: 21',
       'Source', 'Unnamed: 23'],
      dtype='object')

## Malá odbočka - list comprehensions

[Dokumetace](https://docs.python.org/3/tutorial/datastructures.html#list-comprehensions)

Elegantní způsob jak transformovat seznamy.

In [14]:
# Zahřívací příklad: 
# nový seznam obsahuje všechny prvky původního seznamu
my_list = [1,2,3,4,5,6]
[item for item in my_list]

[1, 2, 3, 4, 5, 6]

In [15]:
# Nový seznam obsahuje všechny prvky původního, 
# ale každý prvk je vynásoben 100
[item*100 for item in my_list]

[100, 200, 300, 400, 500, 600]

In [16]:
# Nový seznam obsahuje jenom ty prvky seznamu, které jsou sudé
[item for item in my_list if item%2 == 0]

[2, 4, 6]

In [17]:
# Obdobným způsobem vytvoříme seznam názvu sloupců, 
# které obsahují 'Unnamed' (částečná shoda)
# To jsou sloupce, které chceme vyhodit
col_to_drop = [
    col_name for col_name in unicef.columns if 'Unnamed' in col_name
]
col_to_drop

['Unnamed: 3',
 'Unnamed: 5',
 'Unnamed: 7',
 'Unnamed: 9',
 'Unnamed: 11',
 'Unnamed: 13',
 'Unnamed: 15',
 'Unnamed: 17',
 'Unnamed: 19',
 'Unnamed: 21',
 'Unnamed: 23']

In [18]:
# Odstraníme z tabulky seznam sloupců, 
# které jsme vytvořili v předešlém kroku
unicef.drop(col_to_drop, axis='columns', inplace=True)
unicef.head()

ISO3 Code    Countries    Total     Male   Female    Urban    Rural Poorest  \
0       AFG  Afghanistan     36.3     26.9     46.8     19.7     41.4    40.6   
1       ALB      Albania  7.60232  7.76783  7.41396  6.66002  8.25574  8.8999   
2       DZA      Algeria  2.19084   2.1955  2.18592  1.94815  2.61943  3.4101   
3       AND      Andorra      NaN      NaN      NaN      NaN      NaN     NaN   
4       AGO       Angola     23.7     22.8     24.6     15.1     33.2    37.2   

    Poorer  Middle   Fourth  Richest        Source  
0     46.2    44.4     31.8     16.7      DHS 2015  
1  6.45801  8.8039   6.7463  6.39511   DHS 2008-09  
2  2.57707  1.7089  1.65181  1.42334  MICS 2012-13  
3      NaN     NaN      NaN      NaN           NaN  
4     25.8    22.5     14.4      9.7  IBEP 2008-09

## Jak přejmenuji sloupce?

In [19]:
primates = pd.read_csv('primates.csv')
# Zobrazíme názvy sloupců
primates.columns

Index(['Unnamed: 0', 'Bodywt', 'Brainwt'], dtype='object')

In [20]:
# inplace=True znamená, že se pozmění současná tabulka
primates.rename(
    columns={'Unnamed: 0': 'Species'}, 
    inplace=True
)
primates.head()

Species  Bodywt  Brainwt
0   Potar monkey    10.0      115
1        Gorilla   207.0      406
2          Human    62.0     1320
3  Rhesus monkey     6.8      179
4          Chimp    52.2      440

## Cvičení
 
1. Načtěte data `titanic3.xls` pomocí `read_excel` funkce z pandas
2. Přejmenujte sloupec `pclass` na `passenger_class` a `home.dest` na `home_destination`

In [21]:
titanic = pd.read_excel('titanic3.xls')
titanic.head()

pclass  survived                                             name     sex  \
0       1         1                    Allen, Miss. Elisabeth Walton  female   
1       1         1                   Allison, Master. Hudson Trevor    male   
2       1         0                     Allison, Miss. Helen Loraine  female   
3       1         0             Allison, Mr. Hudson Joshua Creighton    male   
4       1         0  Allison, Mrs. Hudson J C (Bessie Waldo Daniels)  female   

       age  sibsp  parch  ticket      fare    cabin embarked boat   body  \
0  29.0000      0      0   24160  211.3375       B5        S    2    NaN   
1   0.9167      1      2  113781  151.5500  C22 C26        S   11    NaN   
2   2.0000      1      2  113781  151.5500  C22 C26        S  NaN    NaN   
3  30.0000      1      2  113781  151.5500  C22 C26        S  NaN  135.0   
4  25.0000      1      2  113781  151.5500  C22 C26        S  NaN    NaN   

                         home.dest  
0                     St Louis, MO  
1  Montreal, PQ / Chesterville, ON  
2  Montreal, PQ / Chesterville, ON  
3  Montreal, PQ / Chesterville, ON  
4  Montreal, PQ / Chesterville, ON

In [22]:
titanic.columns

Index(['pclass', 'survived', 'name', 'sex', 'age', 'sibsp', 'parch', 'ticket',
       'fare', 'cabin', 'embarked', 'boat', 'body', 'home.dest'],
      dtype='object')

In [23]:
titanic.rename(
    columns={
        'pclass': 'passenger_class', 
        'home.dest':'home_destination'
    },
    inplace=True
)
titanic.head()

passenger_class  survived                                             name  \
0                1         1                    Allen, Miss. Elisabeth Walton   
1                1         1                   Allison, Master. Hudson Trevor   
2                1         0                     Allison, Miss. Helen Loraine   
3                1         0             Allison, Mr. Hudson Joshua Creighton   
4                1         0  Allison, Mrs. Hudson J C (Bessie Waldo Daniels)   

      sex      age  sibsp  parch  ticket      fare    cabin embarked boat  \
0  female  29.0000      0      0   24160  211.3375       B5        S    2   
1    male   0.9167      1      2  113781  151.5500  C22 C26        S   11   
2  female   2.0000      1      2  113781  151.5500  C22 C26        S  NaN   
3    male  30.0000      1      2  113781  151.5500  C22 C26        S  NaN   
4  female  25.0000      1      2  113781  151.5500  C22 C26        S  NaN   

    body                 home_destination  
0    NaN                     St Louis, MO  
1    NaN  Montreal, PQ / Chesterville, ON  
2    NaN  Montreal, PQ / Chesterville, ON  
3  135.0  Montreal, PQ / Chesterville, ON  
4    NaN  Montreal, PQ / Chesterville, ON

## Jak zjistím, v jakém formátu jsou sloupce?

In [24]:
primates.dtypes

Species     object
Bodywt     float64
Brainwt      int64
dtype: object

In [25]:
titanic.dtypes

passenger_class       int64
survived              int64
name                 object
sex                  object
age                 float64
sibsp                 int64
parch                 int64
ticket               object
fare                float64
cabin                object
embarked             object
boat                 object
body                float64
home_destination     object
dtype: object

## Jak zjistím, kolik má tabulká řádů a sloupců?

In [26]:
# Počet řádků a počet sloupců
primates.shape

(5, 3)

In [27]:
# Počet řádků
primates.shape[0]

5

In [28]:
# Počet sloupců
primates.shape[1]

3

In [29]:
# Počet řádků
len(primates)

5

## Cvičení
 
1. Načtěte data `titanic3.xls` pomocí `read_excel` funkce z pandas
2. Zjistěte počet řádků (správná odpověď je 1309)
3. Zjistěte počet sloupců (správná odpověď je 14)

In [30]:
titanic.shape[0]

1309

In [31]:
titanic.shape[1]

14

## Jak zjistím, zda v tabulce jsou chybějící data?

In [32]:
# Načteme data z Excelu
titanic = pd.read_excel('titanic3.xls')
titanic.head()

pclass  survived                                             name     sex  \
0       1         1                    Allen, Miss. Elisabeth Walton  female   
1       1         1                   Allison, Master. Hudson Trevor    male   
2       1         0                     Allison, Miss. Helen Loraine  female   
3       1         0             Allison, Mr. Hudson Joshua Creighton    male   
4       1         0  Allison, Mrs. Hudson J C (Bessie Waldo Daniels)  female   

       age  sibsp  parch  ticket      fare    cabin embarked boat   body  \
0  29.0000      0      0   24160  211.3375       B5        S    2    NaN   
1   0.9167      1      2  113781  151.5500  C22 C26        S   11    NaN   
2   2.0000      1      2  113781  151.5500  C22 C26        S  NaN    NaN   
3  30.0000      1      2  113781  151.5500  C22 C26        S  NaN  135.0   
4  25.0000      1      2  113781  151.5500  C22 C26        S  NaN    NaN   

                         home.dest  
0                     St Louis, MO  
1  Montreal, PQ / Chesterville, ON  
2  Montreal, PQ / Chesterville, ON  
3  Montreal, PQ / Chesterville, ON  
4  Montreal, PQ / Chesterville, ON

In [33]:
# Jaké sloupce má tabulka?
titanic.columns

Index(['pclass', 'survived', 'name', 'sex', 'age', 'sibsp', 'parch', 'ticket',
       'fare', 'cabin', 'embarked', 'boat', 'body', 'home.dest'],
      dtype='object')

In [34]:
# Má tabulká chybějící data?
titanic.isnull().values.any()

True

In [35]:
# Zjistíme, které sloupce mají chybějící data
titanic.isnull().any()

pclass       False
survived     False
name         False
sex          False
age           True
sibsp        False
parch        False
ticket       False
fare          True
cabin         True
embarked      True
boat          True
body          True
home.dest     True
dtype: bool

In [36]:
# Zobrazíme počet chybějících hodnot v každém sloupci
titanic.isnull().sum()

pclass          0
survived        0
name            0
sex             0
age           263
sibsp           0
parch           0
ticket          0
fare            1
cabin        1014
embarked        2
boat          823
body         1188
home.dest     564
dtype: int64

## Cvičení
 
1. Načtěte data `YouthRisk2009.csv`
2. Zjistěte, které sloupce mají chybějící data (správná odpověď Sleep, Sleep7, SmokeLife, SmokeDaily, MarijuaEver)

**Bonus**: Zjistěte kolík řádků má chybějící data (správná odpověď je 72)

In [37]:
youth_risk = pd.read_csv('YouthRisk2009.csv', index_col=0)
youth_risk.head()

Sleep  Sleep7 SmokeLife SmokeDaily  MarijuaEver  Age
1      NaN     NaN        No        NaN          0.0   16
2  8 hours     1.0       Yes        Yes          1.0   16
3  5 hours     0.0       Yes        Yes          1.0   17
4  5 hours     0.0       Yes        Yes          1.0   18
5  7 hours     1.0       Yes         No          1.0   17

In [38]:
# Nastavíme maximální počet řádků, které se zobrazí při náhledu
pd.options.display.max_rows = 100

In [39]:
youth_risk.isnull().any(axis=1)

1       True
2      False
3      False
4      False
5      False
6      False
7      False
8      False
9      False
10     False
11     False
12     False
13     False
14     False
15     False
16     False
17     False
18     False
19     False
20     False
21     False
22     False
23     False
24     False
25     False
26     False
27     False
28     False
29     False
30     False
31     False
32      True
33     False
34     False
35     False
36      True
37     False
38     False
39     False
40     False
41     False
42     False
43     False
44     False
45     False
46     False
47     False
48      True
49     False
50     False
       ...  
451    False
452    False
453    False
454    False
455    False
456    False
457    False
458    False
459    False
460    False
461    False
462     True
463    False
464    False
465    False
466    False
467    False
468    False
469    False
470    False
471    False
472    False
473     True
474    False
475    False
476    False

In [40]:
# Počet řádků s chybějícími hodnotami
youth_risk.isnull().any(axis=1).sum()

72

## Jak vyberu řádky a sloupce?

Pandas má 2 způsoby:
- `iloc` vybírá podle pozic (indexů)
    - stejně jako když vybírám ze seznamu
    - bez koncového indexu
- `loc` vybírá podle jmen
    - včetně koncového indexu

In [41]:
# Načteme data
primates = pd.read_csv('primates.csv')
primates.rename(index=str, columns={'Unnamed: 0': 'Species'}, inplace=True)
primates.head()

Species  Bodywt  Brainwt
0   Potar monkey    10.0      115
1        Gorilla   207.0      406
2          Human    62.0     1320
3  Rhesus monkey     6.8      179
4          Chimp    52.2      440

In [42]:
# Vybereme 0. a 1. řádky
primates.iloc[0:2]

Species  Bodywt  Brainwt
0  Potar monkey    10.0      115
1       Gorilla   207.0      406

In [43]:
# Vybereme 0. a 1. řádky a 1. a 2. sloupce
primates.iloc[0:2, [0,2]]

Species  Brainwt
0  Potar monkey      115
1       Gorilla      406

In [44]:
# Vybíráme podle jména
# POZOR: číselný index musí být převeden řetězec
primates.loc[:'2', ['Species', 'Bodywt']]

Species  Bodywt
0  Potar monkey    10.0
1       Gorilla   207.0
2         Human    62.0

In [45]:
# Vybereme sloupec
primates.Bodywt

0     10.0
1    207.0
2     62.0
3      6.8
4     52.2
Name: Bodywt, dtype: float64

In [46]:
# Vybereme sloupec - druhy způsob
primates['Bodywt']

0     10.0
1    207.0
2     62.0
3      6.8
4     52.2
Name: Bodywt, dtype: float64

In [47]:
# Lze specifikovat, který sloupec je index
# Index se v tomto případě skládá z řetězců
primates2 = pd.read_csv(
    'primates.csv', 
    index_col=0
)
primates2.head()

Bodywt  Brainwt
Potar monkey     10.0      115
Gorilla         207.0      406
Human            62.0     1320
Rhesus monkey     6.8      179
Chimp            52.2      440

In [48]:
primates2.loc['Potar monkey']

Bodywt      10.0
Brainwt    115.0
Name: Potar monkey, dtype: float64

## Cvičení
 
1. Načtěte data `titanic`
2. Vyberte prvních 20 řádků
3. Vyberte prvních 20 řádků a sloupec se jménem (`name`)

**Bonus**: Vyberte řádky 10 až 550 a následující sloupce: jméno (`name`), cena jízdenky (`fare`), zda osoba přežila (`survived`)

In [49]:
titanic.iloc[:20]

pclass  survived                                               name  \
0        1         1                      Allen, Miss. Elisabeth Walton   
1        1         1                     Allison, Master. Hudson Trevor   
2        1         0                       Allison, Miss. Helen Loraine   
3        1         0               Allison, Mr. Hudson Joshua Creighton   
4        1         0    Allison, Mrs. Hudson J C (Bessie Waldo Daniels)   
5        1         1                                Anderson, Mr. Harry   
6        1         1                  Andrews, Miss. Kornelia Theodosia   
7        1         0                             Andrews, Mr. Thomas Jr   
8        1         1      Appleton, Mrs. Edward Dale (Charlotte Lamson)   
9        1         0                            Artagaveytia, Mr. Ramon   
10       1         0                             Astor, Col. John Jacob   
11       1         1  Astor, Mrs. John Jacob (Madeleine Talmadge Force)   
12       1         1                      Aubart, Mme. Leontine Pauline   
13       1         1                       Barber, Miss. Ellen "Nellie"   
14       1         1               Barkworth, Mr. Algernon Henry Wilson   
15       1         0                                Baumann, Mr. John D   
16       1         0                           Baxter, Mr. Quigg Edmond   
17       1         1    Baxter, Mrs. James (Helene DeLaudeniere Chaput)   
18       1         1                              Bazzani, Miss. Albina   
19       1         0                               Beattie, Mr. Thomson   

       sex      age  sibsp  parch    ticket      fare    cabin embarked boat  \
0   female  29.0000      0      0     24160  211.3375       B5        S    2   
1     male   0.9167      1      2    113781  151.5500  C22 C26        S   11   
2   female   2.0000      1      2    113781  151.5500  C22 C26        S  NaN   
3     male  30.0000      1      2    113781  151.5500  C22 C26        S  NaN   
4   female  25.0000      1      2    113781  151.5500  C22 C26        S  NaN   
5     male  48.0000      0      0     19952   26.5500      E12        S    3   
6   female  63.0000      1      0     13502   77.9583       D7        S   10   
7     male  39.0000      0      0    112050    0.0000      A36        S  NaN   
8   female  53.0000      2      0     11769   51.4792     C101        S    D   
9     male  71.0000      0      0  PC 17609   49.5042      NaN        C  NaN   
10    male  47.0000      1      0  PC 17757  227.5250  C62 C64        C  NaN   
11  female  18.0000      1      0  PC 17757  227.5250  C62 C64        C    4   
12  female  24.0000      0      0  PC 17477   69.3000      B35        C    9   
13  female  26.0000      0      0     19877   78.8500      NaN        S    6   
14    male  80.0000      0      0     27042   30.0000      A23        S    B   
15    male      NaN      0      0  PC 17318   25.9250      NaN        S  NaN   
16    male  24.0000      0      1  PC 17558  247.5208  B58 B60        C  NaN   
17  female  50.0000      0      1  PC 17558  247.5208  B58 B60        C    6   
18  female  32.0000      0      0     11813   76.2917      D15        C    8   
19    male  36.0000      0      0     13050   75.2417       C6        C    A   

     body                        home.dest  
0     NaN                     St Louis, MO  
1     NaN  Montreal, PQ / Chesterville, ON  
2     NaN  Montreal, PQ / Chesterville, ON  
3   135.0  Montreal, PQ / Chesterville, ON  
4     NaN  Montreal, PQ / Chesterville, ON  
5     NaN                     New York, NY  
6     NaN                       Hudson, NY  
7     NaN                      Belfast, NI  
8     NaN              Bayside, Queens, NY  
9    22.0              Montevideo, Uruguay  
10  124.0                     New York, NY  
11    NaN                     New York, NY  
12    NaN                    Paris, France  
13    NaN                              NaN  
14    NaN                    Hessle, Yorks  
15    NaN                  

In [50]:
titanic.iloc[:20, 2]

0                         Allen, Miss. Elisabeth Walton
1                        Allison, Master. Hudson Trevor
2                          Allison, Miss. Helen Loraine
3                  Allison, Mr. Hudson Joshua Creighton
4       Allison, Mrs. Hudson J C (Bessie Waldo Daniels)
5                                   Anderson, Mr. Harry
6                     Andrews, Miss. Kornelia Theodosia
7                                Andrews, Mr. Thomas Jr
8         Appleton, Mrs. Edward Dale (Charlotte Lamson)
9                               Artagaveytia, Mr. Ramon
10                               Astor, Col. John Jacob
11    Astor, Mrs. John Jacob (Madeleine Talmadge Force)
12                        Aubart, Mme. Leontine Pauline
13                         Barber, Miss. Ellen "Nellie"
14                 Barkworth, Mr. Algernon Henry Wilson
15                                  Baumann, Mr. John D
16                             Baxter, Mr. Quigg Edmond
17      Baxter, Mrs. James (Helene DeLaudeniere 

In [51]:
titanic.iloc[10:551, [1,2,8]]

survived                                               name      fare
10          0                             Astor, Col. John Jacob  227.5250
11          1  Astor, Mrs. John Jacob (Madeleine Talmadge Force)  227.5250
12          1                      Aubart, Mme. Leontine Pauline   69.3000
13          1                       Barber, Miss. Ellen "Nellie"   78.8500
14          1               Barkworth, Mr. Algernon Henry Wilson   30.0000
15          0                                Baumann, Mr. John D   25.9250
16          0                           Baxter, Mr. Quigg Edmond  247.5208
17          1    Baxter, Mrs. James (Helene DeLaudeniere Chaput)  247.5208
18          1                              Bazzani, Miss. Albina   76.2917
19          0                               Beattie, Mr. Thomson   75.2417
20          1                      Beckwith, Mr. Richard Leonard   52.5542
21          1   Beckwith, Mrs. Richard Leonard (Sallie Monypeny)   52.5542
22          1                              Behr, Mr. Karl Howell   30.0000
23          1                              Bidois, Miss. Rosalie  227.5250
24          1                                  Bird, Miss. Ellen  221.7792
25          0                                Birnbaum, Mr. Jakob   26.0000
26          1                            Bishop, Mr. Dickinson H   91.0792
27          1            Bishop, Mrs. Dickinson H (Helen Walton)   91.0792
28          1                             Bissette, Miss. Amelia  135.6333
29          1          Bjornstrom-Steffansson, Mr. Mauritz Hakan   26.5500
30          0                       Blackwell, Mr. Stephen Weart   35.5000
31          1                                   Blank, Mr. Henry   31.0000
32          1                            Bonnell, Miss. Caroline  164.8667
33          1                           Bonnell, Miss. Elizabeth   26.5500
34          0                           Borebank, Mr. John James   26.5500
35          1                           Bowen, Miss. Grace Scott  262.3750
36          1                        Bowerman, Miss. Elsie Edith   55.0000
37          1      Bradley, Mr. George ("George Arthur Brayton")   26.5500
38          0                            Brady, Mr. John Bertram   30.5000
39          0                                 Brandeis, Mr. Emil   50.4958
40          0                          Brewe, Dr. Arthur Jackson   39.6000
41          1          Brown, Mrs. James Joseph (Margaret Tobin)   27.7208
42          1     Brown, Mrs. John Murray (Caroline Lane Lamson)   51.4792
43          1    Bucknell, Mrs. William Robert (Emma Eliza Ward)   76.2917
44          1                    Burns, Miss. Elizabeth Margaret  134.5000
45          0                  Butt, Major. Archibald Willingham   26.5500
46          0                              Cairns, Mr. Alexander   31.0000
47          1                  Calderhead, Mr. Edward Pennington   26.2875
48          1   Candee, Mrs. Edward (Helen Churchill Hungerford)   27.4458
49          1                 Cardeza, Mr. Thomas Drake Martinez  512.3292
50          1  Cardeza, Mrs. James Warburton Martinez (Charlo...  512.3292
51          0                           Carlsson, Mr. Frans Olof    5.0000
52          0                            Carrau, Mr. Francisco M   47.1000
53          0                             Carrau, Mr. Jose Pedro   47.1000
54          1                Carter, Master. William Thornton II  120.0000
55          1                          Carter, Miss. Lucile Polk  120.0000
56          1                         Carter, Mr. William Ernest  120.0000
57          1          Carter, Mrs. William Ernest (Lucile Polk)  120.0000
58          0                             Case, Mr. Howard Brown   26.0000
59          1  Cassebeer, Mrs. Henry Arthur Jr (Eleanor Genev...   27.7208
..        ...                                                ...       ...
501         1                  Mellinger, Miss. Madeleine Violet   19.5000
502         1          Melling

In [52]:
titanic.loc[10:550, ['name', 'fare', 'survived']]

name      fare  survived
10                              Astor, Col. John Jacob  227.5250         0
11   Astor, Mrs. John Jacob (Madeleine Talmadge Force)  227.5250         1
12                       Aubart, Mme. Leontine Pauline   69.3000         1
13                        Barber, Miss. Ellen "Nellie"   78.8500         1
14                Barkworth, Mr. Algernon Henry Wilson   30.0000         1
15                                 Baumann, Mr. John D   25.9250         0
16                            Baxter, Mr. Quigg Edmond  247.5208         0
17     Baxter, Mrs. James (Helene DeLaudeniere Chaput)  247.5208         1
18                               Bazzani, Miss. Albina   76.2917         1
19                                Beattie, Mr. Thomson   75.2417         0
20                       Beckwith, Mr. Richard Leonard   52.5542         1
21    Beckwith, Mrs. Richard Leonard (Sallie Monypeny)   52.5542         1
22                               Behr, Mr. Karl Howell   30.0000         1
23                               Bidois, Miss. Rosalie  227.5250         1
24                                   Bird, Miss. Ellen  221.7792         1
25                                 Birnbaum, Mr. Jakob   26.0000         0
26                             Bishop, Mr. Dickinson H   91.0792         1
27             Bishop, Mrs. Dickinson H (Helen Walton)   91.0792         1
28                              Bissette, Miss. Amelia  135.6333         1
29           Bjornstrom-Steffansson, Mr. Mauritz Hakan   26.5500         1
30                        Blackwell, Mr. Stephen Weart   35.5000         0
31                                    Blank, Mr. Henry   31.0000         1
32                             Bonnell, Miss. Caroline  164.8667         1
33                            Bonnell, Miss. Elizabeth   26.5500         1
34                            Borebank, Mr. John James   26.5500         0
35                            Bowen, Miss. Grace Scott  262.3750         1
36                         Bowerman, Miss. Elsie Edith   55.0000         1
37       Bradley, Mr. George ("George Arthur Brayton")   26.5500         1
38                             Brady, Mr. John Bertram   30.5000         0
39                                  Brandeis, Mr. Emil   50.4958         0
40                           Brewe, Dr. Arthur Jackson   39.6000         0
41           Brown, Mrs. James Joseph (Margaret Tobin)   27.7208         1
42      Brown, Mrs. John Murray (Caroline Lane Lamson)   51.4792         1
43     Bucknell, Mrs. William Robert (Emma Eliza Ward)   76.2917         1
44                     Burns, Miss. Elizabeth Margaret  134.5000         1
45                   Butt, Major. Archibald Willingham   26.5500         0
46                               Cairns, Mr. Alexander   31.0000         0
47                   Calderhead, Mr. Edward Pennington   26.2875         1
48    Candee, Mrs. Edward (Helen Churchill Hungerford)   27.4458         1
49                  Cardeza, Mr. Thomas Drake Martinez  512.3292         1
50   Cardeza, Mrs. James Warburton Martinez (Charlo...  512.3292         1
51                            Carlsson, Mr. Frans Olof    5.0000         0
52                             Carrau, Mr. Francisco M   47.1000         0
53                              Carrau, Mr. Jose Pedro   47.1000         0
54                 Carter, Master. William Thornton II  120.0000         1
55                           Carter, Miss. Lucile Polk  120.0000         1
56                          Carter, Mr. William Ernest  120.0000         1
57           Carter, Mrs. William Ernest (Lucile Polk)  120.0000         1
58                              Case, Mr. Howard Brown   26.0000         0
59   Cassebeer, Mrs. Henry Arthur Jr (Eleanor Genev...   27.7208         1
..                                                 ...       ...       ...
501                  Mellinger, Miss. Madeleine Violet   19.5000         1
502          Mellinger, Mrs. (Elizabeth Anne Maidment)   19.5000         1


## Počítání s celými sloupci

* Pandas umí počítat s celým sloupcem najednou

Primáti:

* Jaká je hmotnost mozku primátů v kilogramech?
* Jaký je poměr mezi hmotností mozku a těla?

In [53]:
primates

Species  Bodywt  Brainwt
0   Potar monkey    10.0      115
1        Gorilla   207.0      406
2          Human    62.0     1320
3  Rhesus monkey     6.8      179
4          Chimp    52.2      440

In [54]:
brainwt_kg = primates.Brainwt / 1000
brainwt_kg

0    0.115
1    0.406
2    1.320
3    0.179
4    0.440
Name: Brainwt, dtype: float64

In [55]:
brainwt_kg / primates.Bodywt

0    0.011500
1    0.001961
2    0.021290
3    0.026324
4    0.008429
dtype: float64

In [56]:
# První způsob vytvoření nového sloupce
primates['ratio'] = brainwt_kg / primates.Bodywt
primates

Species  Bodywt  Brainwt     ratio
0   Potar monkey    10.0      115  0.011500
1        Gorilla   207.0      406  0.001961
2          Human    62.0     1320  0.021290
3  Rhesus monkey     6.8      179  0.026324
4          Chimp    52.2      440  0.008429

In [57]:
# První způsob vytvoření nového sloupce
primates.assign(ratio2=(brainwt_kg / primates.Bodywt), inplace=True)
primates

Species  Bodywt  Brainwt     ratio
0   Potar monkey    10.0      115  0.011500
1        Gorilla   207.0      406  0.001961
2          Human    62.0     1320  0.021290
3  Rhesus monkey     6.8      179  0.026324
4          Chimp    52.2      440  0.008429

In [1]:
# Nevektorové oprerace - iterujeme řádky tabulky
for row_index, row in primates.iterrows():
    print(row)

NameError: name 'primates' is not defined

## Jak vyberu řádky, které splňují určitou podmínku?

`data[podmínka]`

`data[(podmínka1) & (podmínka2)]`

In [59]:
primates2

Bodywt  Brainwt
Potar monkey     10.0      115
Gorilla         207.0      406
Human            62.0     1320
Rhesus monkey     6.8      179
Chimp            52.2      440

In [60]:
primates2.columns

Index(['Bodywt', 'Brainwt'], dtype='object')

In [61]:
primates2.Bodywt

Potar monkey      10.0
Gorilla          207.0
Human             62.0
Rhesus monkey      6.8
Chimp             52.2
Name: Bodywt, dtype: float64

In [62]:
primates2.Bodywt > 60

Potar monkey     False
Gorilla           True
Human             True
Rhesus monkey    False
Chimp            False
Name: Bodywt, dtype: bool

In [63]:
#  ~ znamená negace
~(primates2.Brainwt < 500)

Potar monkey     False
Gorilla          False
Human             True
Rhesus monkey    False
Chimp            False
Name: Brainwt, dtype: bool

In [64]:
# | znamená nebo
primates2[(primates2.Bodywt > 60) | (primates2.Brainwt >= 500)]

Bodywt  Brainwt
Gorilla   207.0      406
Human      62.0     1320

In [65]:
# Vyberéme řádky, kde tělesná hmota je více než 60
primates2[primates2.Bodywt > 60]

Bodywt  Brainwt
Gorilla   207.0      406
Human      62.0     1320

In [66]:
# Vyberéme řádky, kde je tělesná hmota více než 60 a
# hmota mozku je méně než 500
primates2[(primates2.Bodywt > 60) & (primates2.Brainwt < 500)]

Bodywt  Brainwt
Gorilla   207.0      406

 ## Cvičení
 
1. Načtěte data `titanic`
2. Vyberte řádky cestujících ženského pohlaví a spočítejte je (správná odpověď je 466)
3. Vyberte řádky cestujících ženského pohlaví, které jsou starší 1 roku a spočítejte je (správná odpověď je 379)

**Bonus**: Pro vybrané řádky spočítejte průměrnou cenu jízdenky (`fare`) (správná odpověď je cca 51)

In [67]:
sum(titanic.sex == 'female')

466

In [68]:
len(titanic[titanic.sex == 'female'])

466

In [69]:
titanic[titanic.sex == 'female'].shape[0]

466

In [70]:
sum((titanic.age > 1) & (titanic.sex == 'female'))

379

In [71]:
titanic[(titanic.age > 1) & (titanic.sex == 'female')].fare.mean()

51.017481530343005

## Jak seskupím řádky?

Použijeme metodu `groupby`, která seskupí řádky dle jednoho nebo více sloupců.

Potom použijeme nějakou metodu nad seskupenými řádky (`count`, `mean`, `sum`).

In [72]:
# Načteme data z Excelu
data = pd.read_excel('titanic3.xls')
data.head()

pclass  survived                                             name     sex  \
0       1         1                    Allen, Miss. Elisabeth Walton  female   
1       1         1                   Allison, Master. Hudson Trevor    male   
2       1         0                     Allison, Miss. Helen Loraine  female   
3       1         0             Allison, Mr. Hudson Joshua Creighton    male   
4       1         0  Allison, Mrs. Hudson J C (Bessie Waldo Daniels)  female   

       age  sibsp  parch  ticket      fare    cabin embarked boat   body  \
0  29.0000      0      0   24160  211.3375       B5        S    2    NaN   
1   0.9167      1      2  113781  151.5500  C22 C26        S   11    NaN   
2   2.0000      1      2  113781  151.5500  C22 C26        S  NaN    NaN   
3  30.0000      1      2  113781  151.5500  C22 C26        S  NaN  135.0   
4  25.0000      1      2  113781  151.5500  C22 C26        S  NaN    NaN   

                         home.dest  
0                     St Louis, MO  
1  Montreal, PQ / Chesterville, ON  
2  Montreal, PQ / Chesterville, ON  
3  Montreal, PQ / Chesterville, ON  
4  Montreal, PQ / Chesterville, ON

In [73]:
# Groupby vrací iterátor
groups = titanic.groupby('sex')

In [74]:
# Unikátní hodnoty dle kterých je rozdělena tabulka
groups.groups.keys()

dict_keys(['female', 'male'])

In [75]:
# Seskupíme řádky dle pohlaví a spočítáme průměrnou hodnotu všech číselných proměnných
data.groupby('sex').mean()

pclass  survived        age     sibsp     parch       fare  \
sex                                                                    
female  2.154506  0.727468  28.687071  0.652361  0.633047  46.198097   
male    2.372479  0.190985  30.585233  0.413998  0.247924  26.154601   

             body  
sex                
female  166.62500  
male    160.39823

In [76]:
# Totéž můžeme udělat pomocí filtrování
titanic[titanic.sex == 'female'].mean()

pclass        2.154506
survived      0.727468
age          28.687071
sibsp         0.652361
parch         0.633047
fare         46.198097
body        166.625000
dtype: float64

In [77]:
titanic[titanic.sex == 'male'].mean()

pclass        2.372479
survived      0.190985
age          30.585233
sibsp         0.413998
parch         0.247924
fare         26.154601
body        160.398230
dtype: float64

In [78]:
# Ve smyčce vytiskneme průměrné hodnoty pro cestující obou pohlaví
for sex in titanic.sex.unique():
    print(sex, titanic[titanic.sex == sex].mean())

female pclass        2.154506
survived      0.727468
age          28.687071
sibsp         0.652361
parch         0.633047
fare         46.198097
body        166.625000
dtype: float64
male pclass        2.372479
survived      0.190985
age          30.585233
sibsp         0.413998
parch         0.247924
fare         26.154601
body        160.398230
dtype: float64


In [79]:
# Seskupení dle více proměnných - pohlaví a místa nalodění
titanic.groupby(['sex', 'embarked', 'survived']).mean()

pclass        age     sibsp     parch       fare  \
sex    embarked survived                                                       
female C        0         2.636364  27.071429  0.454545  0.454545  17.460991   
                1         1.539216  31.543956  0.509804  0.519608  87.994648   
       Q        0         3.000000  26.692308  0.086957  0.347826   9.311961   
                1         2.837838  24.000000  0.270270  0.000000  14.564192   
       S        0         2.806452  24.842105  1.333333  1.075269  26.117341   
                1         1.924242  29.104167  0.560606  0.651515  45.549622   
male   C        0         2.119266  35.178082  0.302752  0.201835  42.555965   
                1         1.729167  29.912602  0.375000  0.416667  63.013896   
       Q        0         2.875000  32.104167  0.500000  0.107143  12.561384   
                1         3.000000  25.000000  0.285714  0.000000   9.974400   
       S        0         2.460348  30.855829  0.444874  0.241779  20.722284   
                1         2.188679  25.712862  0.358491  0.339623  27.291978   

                                body  
sex    embarked survived              
female C        0         328.000000  
                1                NaN  
       Q        0         229.000000  
                1                NaN  
       S        0          79.500000  
                1                NaN  
male   C        0         155.083333  
                1                NaN  
       Q        0         134.750000  
                1                NaN  
       S        0         163.105882  
                1                NaN

 ## Cvičení

Seskupte `titanic` dle třídy (`pclass`) a pohlaví (`sex`) zobrazte medián (`median`)

**Bonus**: zobrazte jenom sloupec věk (`age`)

In [80]:
titanic.groupby(['pclass', 'sex']).median()

survived   age  sibsp  parch      fare   body
pclass sex                                                  
1      female         1  36.0      0      0  80.92915    NaN
       male           0  42.0      0      0  49.50420  166.0
2      female         1  28.0      0      0  23.00000   52.0
       male           0  29.5      0      0  13.00000  160.0
3      female         0  22.0      0      0  10.48960  206.0
       male           0  25.0      0      0   7.89580  148.0

In [81]:
# Pokud seskupojeme podle 2 a více proměnných, výsledná tabulka 
# bude mít několikaúrovňový index
titanic.groupby(['pclass', 'sex']).median()['age'].index

MultiIndex(levels=[[1, 2, 3], ['female', 'male']],
           labels=[[0, 0, 1, 1, 2, 2], [0, 1, 0, 1, 0, 1]],
           names=['pclass', 'sex'])

In [82]:
titanic.groupby(['pclass',]).median()['age'].index

Int64Index([1, 2, 3], dtype='int64', name='pclass')

In [83]:
titanic.groupby(['pclass', 'sex']).median().age

pclass  sex   
1       female    36.0
        male      42.0
2       female    28.0
        male      29.5
3       female    22.0
        male      25.0
Name: age, dtype: float64

## Jak ukážu popisnou statistiku pro všechny sloupce?

In [84]:
# Bez include='all' jenom numerické sloupce
titanic.describe(include='all')

pclass     survived                  name   sex          age  \
count   1309.000000  1309.000000                  1309  1309  1046.000000   
unique          NaN          NaN                  1307     2          NaN   
top             NaN          NaN  Connolly, Miss. Kate  male          NaN   
freq            NaN          NaN                     2   843          NaN   
mean       2.294882     0.381971                   NaN   NaN    29.881135   
std        0.837836     0.486055                   NaN   NaN    14.413500   
min        1.000000     0.000000                   NaN   NaN     0.166700   
25%        2.000000     0.000000                   NaN   NaN    21.000000   
50%        3.000000     0.000000                   NaN   NaN    28.000000   
75%        3.000000     1.000000                   NaN   NaN    39.000000   
max        3.000000     1.000000                   NaN   NaN    80.000000   

              sibsp        parch    ticket         fare        cabin embarked  \
count   1309.000000  1309.000000      1309  1308.000000          295     1307   
unique          NaN          NaN       939          NaN          186        3   
top             NaN          NaN  CA. 2343          NaN  C23 C25 C27        S   
freq            NaN          NaN        11          NaN            6      914   
mean       0.498854     0.385027       NaN    33.295479          NaN      NaN   
std        1.041658     0.865560       NaN    51.758668          NaN      NaN   
min        0.000000     0.000000       NaN     0.000000          NaN      NaN   
25%        0.000000     0.000000       NaN     7.895800          NaN      NaN   
50%        0.000000     0.000000       NaN    14.454200          NaN      NaN   
75%        1.000000     0.000000       NaN    31.275000          NaN      NaN   
max        8.000000     9.000000       NaN   512.329200          NaN      NaN   

       boat        body     home.dest  
count   486  121.000000           745  
unique   28         NaN           369  
top      13         NaN  New York, NY  
freq     39         NaN            64  
mean    NaN  160.809917           NaN  
std     NaN   97.696922           NaN  
min     NaN    1.000000           NaN  
25%     NaN   72.000000           NaN  
50%     NaN  155.000000           NaN  
75%     NaN  256.000000           NaN  
max     NaN  328.000000           NaN

In [85]:
# Bez include='all' jenom numerické sloupce
titanic.describe()

pclass     survived          age        sibsp        parch  \
count  1309.000000  1309.000000  1046.000000  1309.000000  1309.000000   
mean      2.294882     0.381971    29.881135     0.498854     0.385027   
std       0.837836     0.486055    14.413500     1.041658     0.865560   
min       1.000000     0.000000     0.166700     0.000000     0.000000   
25%       2.000000     0.000000    21.000000     0.000000     0.000000   
50%       3.000000     0.000000    28.000000     0.000000     0.000000   
75%       3.000000     1.000000    39.000000     1.000000     0.000000   
max       3.000000     1.000000    80.000000     8.000000     9.000000   

              fare        body  
count  1308.000000  121.000000  
mean     33.295479  160.809917  
std      51.758668   97.696922  
min       0.000000    1.000000  
25%       7.895800   72.000000  
50%      14.454200  155.000000  
75%      31.275000  256.000000  
max     512.329200  328.000000

## Cvičení

Použijte následující postup a seznamte se s [těmito daty](http://vincentarelbundock.github.io/Rdatasets/csv/robustbase/education.csv), [popis dat](http://vincentarelbundock.github.io/Rdatasets/doc/robustbase/education.html)


1. Načíst data
    - Případně přejmenovat sloupce

2. V jakém formátu jsou sloupce?

3. Kolik je řádků a sloupců?

4. Obsahují některé sloupce chybějící data?

5. Udělat popisnou statistiku dat

### Načíst data

In [86]:
education = pd.read_csv(
    'http://vincentarelbundock.github.io/Rdatasets/csv/robustbase/education.csv',
    index_col=0
)
education.head()

State  Region   X1    X2   X3    Y
1    ME       1  508  3944  325  235
2    NH       1  564  4578  323  231
3    VT       1  322  4011  328  270
4    MA       1  846  5233  305  261
5    RI       1  871  4780  303  300

### Přejmenovat sloupce

In [87]:
education.rename(columns={
    'X1': 'residents', 'X2': 'income', 
    'X3': 'residents_young', 'Y': 'expenditure'
    }, 
    inplace=True
)
education.head()

State  Region  residents  income  residents_young  expenditure
1    ME       1        508    3944              325          235
2    NH       1        564    4578              323          231
3    VT       1        322    4011              328          270
4    MA       1        846    5233              305          261
5    RI       1        871    4780              303          300

### V jakém formátu jsou sloupce?

In [88]:
education.dtypes

State              object
Region              int64
residents           int64
income              int64
residents_young     int64
expenditure         int64
dtype: object

### Kolik je řádků a sloupců?

In [89]:
education.shape

(50, 6)

### Obsahují některé sloupce chybějící data?

In [90]:
education.isnull().any()

State              False
Region             False
residents          False
income             False
residents_young    False
expenditure        False
dtype: bool

### Udělat popisnou statistiku dat

In [91]:
education.describe(include='all')

State     Region   residents       income  residents_young  expenditure
count     50  50.000000   50.000000    50.000000        50.000000    50.000000
unique    50        NaN         NaN          NaN              NaN          NaN
top       CO        NaN         NaN          NaN              NaN          NaN
freq       1        NaN         NaN          NaN              NaN          NaN
mean     NaN   2.660000  657.800000  4675.120000       325.740000   284.600000
std      NaN   1.061574  145.016396   644.506254        19.423119    61.340136
min      NaN   1.000000  322.000000  3448.000000       287.000000   208.000000
25%      NaN   2.000000  546.750000  4137.250000       310.750000   234.250000
50%      NaN   3.000000  662.500000  4706.000000       324.500000   269.500000
75%      NaN   3.750000  782.250000  5054.250000       333.000000   316.750000
max      NaN   4.000000  909.000000  5889.000000       386.000000   546.000000